### PR TITLE
Visual test that covers DisplayObject, SystemRenderer, Extract.

### PR DESCRIPTION
**Pull request by ivanpopelyshev**
Update 20/02/2018: It is WIP again, I've just found a way to make `generateTexture` and `extract` results more consistent, and clean up the code.

## Overview

I was going to add a few params in Extract API, and made a test. I dont like to make tests, but when I do, my test breaks several subsystems and leads to refactoring.

I can split to several individual tests and rework API's in v5.

That PR was completed in friday night.

### :gift: Added

* `renderer.extract.data` accepts `region` and `resolution` parameters, returns `CanvasData` instant that containss pixel data, their format, region and resolution it was taken from
* Adds `CanvasEquals` tests module that takes into account flipY and premultiplyAlpha problems
* Adds megatest that covers several subsystems

### ✏️ Changed

* `generateTexture` API, new `region` parameter replaces `scaleMode`, which is now the final parameter.
* All temporary parent/transform operations are moved in `UpdateTransformHelper` class.

### 🐛 Fixed

* All temporary changes of parent and transform in `DisplayObject`, now take into account several cases of lazy displayObject evaluations.
* `CanvasRenderer` and `WebGLRenderer` use `UpdateTransformHelper` API and don't participate in those hacks.

### 🔎 Solved issues

#2695
#4624
#4530
#4670, [fiddle](https://jsfiddle.net/ASCII_HUSKY/7dt7fpsv/27/)
#3850, resolution can be passed through new function

### CODE REVIEW

**Comment by ivanpopelyshev** [comment]
@setpixel try this build: pixijs.download/dev-extract-params/pixi.js , submit "true" as third parameter to canvas/pixels.

**Comment by bigtimebuddy** [comment]
I think we need some unit-tests here, especially for `flipY`, `postDivide` and support for the `region` param.

**Comment by ivanpopelyshev** [comment]
Yep, lets wait when someone has free time for this thing. 

I made it to address all those issues, its nor critical nor high priority ;)

**Comment by ivanpopelyshev** [comment]
Now we have a class that combines data, dimensions and flags (flip, premultiply).

In v5 its better to just add `extract` function to renderer as a mixin that returns it.

I've added a test but it is failing :(

File: test/extract/index.js [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+const EqualsData = require('./equals');
+
+describe('PIXI.extract', function ()
+{
+    it('should use generateTexture, check the region, premultiplyAlpha and flipY', function (done)
+    {
+        const renderers = [new PIXI.CanvasRenderer(36, 47, { transparent: true })];
+
+        // renderers.pop();
```
Comment (cursedcoder) [inline test/extract/index.js]: remove?

File: test/extract/equals.js [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+/**
+ * equality that takes into account flips and multiplied alpha
+ * @param {PIXI.extract.CanvasData} arg1 the other canvas data
+ * @param {PIXI.extract.CanvasData} arg2 the other canvas data
+ * @returns {boolean} result
+ */
+
+/* eslint-disable no-console */
+module.exports = function equals(arg1, arg2)
```
Comment (cursedcoder) [inline test/extract/equals.js]: `export default function`

File: test/extract/equals.js [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+            alpha2 = arg2.premultiplyAlpha ? 1.0 : (alpha2 / 255.0);
+
+            for (let t = 0; t < 3; t++)
+            {
+                const R1 = Math.round(pixels1[ind1] * alpha1);
+                const R2 = Math.round(pixels2[ind2] * alpha2);
+
+                if (Math.abs(R1 - R2) > 1)
+                {
+                    console.log(`Image comparsion: color fail x=${x} y=${y}, a1=${alpha1}, a2=${alpha2},
```
Comment (cursedcoder) [inline test/extract/equals.js]: Image compa*_riso_*n

File: test/extract/equals.js [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+                return false;
+            }
+
+            alpha1 = arg1.premultiplyAlpha ? 1.0 : (alpha1 / 255.0);
+            alpha2 = arg2.premultiplyAlpha ? 1.0 : (alpha2 / 255.0);
+
+            for (let t = 0; t < 3; t++)
+            {
+                const R1 = Math.round(pixels1[ind1] * alpha1);
```
Comment (cursedcoder) [inline test/extract/equals.js]: What does `R` mean here?

File: src/extract/webgl/WebGLExtract.js [hunk 5]
```diff
@@ -8,10 +9,12 @@ const BYTES_PER_PIXEL = 4;
  *
  * An instance of this class is automatically created by default, and can be found at renderer.plugins.extract
  *
+ * All methods use global coordinates of the target. Please assign temporary parent and call updateTransform to use local.
+ *
  * @class
  * @memberof PIXI.extract
  */
-export default class WebGLExtract
```
Comment (cursedcoder) [inline src/extract/webgl/WebGLExtract.js]: Why remove?

File: src/extract/CanvasData.js [hunk 6]
```diff
@@ -0,0 +1,205 @@
+import * as core from '../core';
+
+/**
+ * Contains extracted data
+ * @class
+ * @memberof PIXI.extract
+ */
+export default class CanvasData
```
Comment (cursedcoder) [inline src/extract/CanvasData.js]: I like it, +1

File: test/extract/equals.js (hunk 2 above)
Comment (ivanpopelyshev) [reply test/extract/equals.js]: Can't, tests are old node modules.

File: test/extract/equals.js (hunk 4 above)
Comment (ivanpopelyshev) [reply test/extract/equals.js]: Premultiplied color.

**Comment by ivanpopelyshev** [comment]
I can make temporary displayObject and Transform global. If something taints them then whole pixi will be unusable, not only one object.

`copy` fields store real values of `parent` and `transform`, one is temporary transform for `getLocalBounds` made the same way temporary parent. One flag to store the type of operation performed on the element. 

I guess we can move it to one big temporary object to reduce number of fields.

The problem with those functions is that I cant reduce them - everything less will fail my megatest in strange ways, for example when we call second `renderer.render` it can break because of `wrongID` weren't updated properly when doing those operations.

**Comment by bigtimebuddy** [comment]
Hey @ivanpopelyshev: no refactors for v4, sorry. Only new features for v4 if they are extremely high value. This is not, IMO, and would be "nice to have". Will _only_ consider the API additions for extract in v5.

Your bug fixes are useful, but please do it one PR at a time. It's too difficult to verify things are fixed when you do it all together and it's more likely to cause instability issues.

**Pull request closed without merging**